src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1111 @@
+    UA_FileContext *fileContext = getFileContext(fileInfo, sessionId, fileHandle);
+
+    if(!fileContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    if(fileContext->openFileMode & UA_OPENFILEMODE_READ)
+        return UA_STATUSCODE_BADINVALIDSTATE;
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+     UA_StatusCode retval =
+         UA_decodeBinary(&data, &trustList, &UA_TYPES[UA_TYPES_TRUSTLISTDATATYPE], NULL);
+
+    if(retval != UA_STATUSCODE_GOOD)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_CertificateGroup *transactionCertGroup =
+        UA_GDSTransaction_getCertificateGroup(server->transaction, certGroup);
+    if(!transactionCertGroup)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    return transactionCertGroup->setTrustList(transactionCertGroup, &trustList);
+}
+
+static UA_StatusCode
+closeTrustList(UA_Server *server,
+               const UA_NodeId *sessionId, void *sessionHandle,
+               const UA_NodeId *methodId, void *methodContext,
+               const UA_NodeId *objectId, void *objectContext,
+               size_t inputSize, const UA_Variant *input,
+               size_t outputSize, UA_Variant *output) {
+    /*check for input types*/
+    if(!UA_Variant_hasScalarType(&input[0], &UA_TYPES[UA_TYPES_UINT32])) /*FileHandle*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_UInt32 fileHandle = *(UA_UInt32*)input[0].data;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    if(!certGroup->fileInfoContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_FileInfo *fileInfo = (UA_FileInfo*)certGroup->fileInfoContext;
+    UA_FileContext *fileContext = getFileContext(fileInfo, sessionId, fileHandle);
+
+    if(!fileContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    /* TODO: needs to be change! */
+    if(fileContext->openFileMode & (UA_OPENFILEMODE_WRITE | UA_OPENFILEMODE_ERASEEXISTING)) {
+        if(server->transaction) {
+            UA_GDSTransaction_clear(server->transaction);
+            server->transaction = NULL;
+        }
+    }
+
+    LIST_REMOVE(fileContext, listEntry);
+    fileInfo->openCount -= 1;
+
+    UA_ByteString_delete(fileContext->file);
+    UA_free(fileContext);
+    if(fileInfo->openCount == 0) {
+        UA_free(fileInfo);
+        certGroup->fileInfoContext = NULL;
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+closeAndUpdateTrustList(UA_Server *server,
+               const UA_NodeId *sessionId, void *sessionHandle,
+               const UA_NodeId *methodId, void *methodContext,
+               const UA_NodeId *objectId, void *objectContext,
+               size_t inputSize, const UA_Variant *input,
+               size_t outputSize, UA_Variant *output) {
+    UA_StatusCode retval = UA_Server_applyChanges(server);
```

Comment:
As stated above: CloseAndUpdate initiates a new transaction (or appends to an existing one) but does not end a transaction.